Issue #108: moved subscribe logic from pub to sub and xsub sockets.

stitchinthyme opened the pull request:
Also added a subscribe all in proxy.go, because otherwise it won't work with pub/sub sockets. This has no effect for other socket types.

sbinet commented on xsub.go:
```diff
@@ -7,18 +7,24 @@ package zmq4
 // The returned socket value is initially unbound.
 func NewXSub(ctx context.Context, opts ...Option) Socket {
-	xsub := &xsubSocket{newSocket(ctx, XSub, opts...)}
+	xsub := &xsubSocket{newSocket(ctx, XSub, opts...), sync.RWMutex{}, make(map[string]struct{})}
```
```suggestion
	xsub := &xsubSocket{sck: newSocket(ctx, XSub, opts...)}
```

or:
```suggestion
	xsub := &xsubSocket{
		sck:    newSocket(ctx, XSub, opts...),
		topics: make(map[string]struct{}),
	}
```
(and drop the extra `make(map[...])` of line `19`.)

stitchinthyme replied:
Committed in branch issue-108.

sbinet commented on xsub.go:
```diff
@@ -7,18 +7,24 @@ package zmq4
 // xsubSocket is a XSUB ZeroMQ socket.
 type xsubSocket struct {
 	sck *socket
+	mu     sync.RWMutex
```
please insert a blank line above `mu` to clearly delimitate the block that mutex is supposed to guard.

stitchinthyme replied:
Committed in issue-108.

sbinet commented on proxy.go:
```diff
@@ -102,6 +102,13 @@ func (p *Proxy) init(front, back, capture Socket) {
+	// Subscribe all. If we don't do this, the sub socket will drop all messages.
+	//  It has no effect for other socket types.
+	if err := front.SetOption(OptionSubscribe, ""); err != nil {
+		log.Fatalf("Subscribe error: %v\n", err)
```
perhaps return an error instead of fatal-ing?
this would involve changing the signature of `Proxy.init` and `NewProxy`... but ok.

stitchinthyme replied:
Updated in branch issue-108.

sbinet commented on pub.go:
```diff
@@ -11,12 +11,6 @@ import (
 	"sync"
 )
 
-// Topics is an interface that wraps the basic Topics method.
```
why do you remove it altogether?
(instead of just moving it to `sub.go`?)

it seems to me it's still a rather interesting interface (one that `Sub` still implements in your new scheme).
WDYT?

stitchinthyme replied:
Makes sense. Moved to sub.go.

sbinet commented on sub.go:
```diff
@@ -138,7 +145,24 @@ func (sub *subSocket) subscribe(topic string, v int) {
 var (
 	_ Socket = (*subSocket)(nil)
-	_ Topics = (*subSocket)(nil)
+	//_ Topics = (*subSocket)(nil)
```
I'd rather keep this interface "contract".

stitchinthyme replied:
Ok, committed in issue-108.

sbinet commented on xsub.go:
```diff
@@ -72,7 +88,64 @@ func (xsub *xsubSocket) GetOption(name string) (interface{}, error) {
+
+	switch name {
+	case OptionSubscribe:
+		k := value.(string)
+		xsub.subscribe(k, 1)
+		topic = append([]byte{1}, k...)
+
+	case OptionUnsubscribe:
+		k := value.(string)
+		topic = append([]byte{0}, k...)
+		xsub.subscribe(k, 0)
```
any reason why in one case it's `subscribe` then `append` while in the other case it's the reverse?

stitchinthyme replied:
Just an oversight. Committed change in issue-108.

sbinet commented on xsub.go:
```diff
@@ -72,7 +88,64 @@ func (xsub *xsubSocket) GetOption(name string) (interface{}, error) {
+		return ErrBadProperty
+	}
+
+	xsub.sck.mu.RLock()
```
could you add a comment on why we take the lock here?

stitchinthyme replied:
I copied this bit from sub.go in master - see line 110. I do not know why it was there in the first place; I was just copying the existing subscribe logic, and didn't want to change anything that didn't seem necessary.

sbinet replied:
ok.
seems like I'd need somebody else to review _my_ PRs :)

sbinet reviewed (commented):
apologies for the belated review.

I have a couple of cosmetics comments, see below.

also: could you add an example that shows the new semantics? (or a test?)

thanks a lot.

sbinet commented on proxy.go:
```diff
@@ -40,15 +40,18 @@ const (
 	}
-	proxy.init(front, back, capture)
-	return &proxy
+	if err := proxy.init(front, back, capture); err != nil {
+		return nil, err
+	} else {
+		return nil, err
+	}
```
this should probably read instead:
```suggestion
	if err := proxy.init(front, back, capture); err != nil {
		return nil, err
	}
	return proxy, nil
```

stitchinthyme replied:
That's what happens when I work too quickly. Committed.

sbinet commented on proxy_test.go:
```diff
@@ -260,7 +260,7 @@ func TestProxy(t *testing.T) {
 	grp.Go(func() error {
 		wg1.Wait() // sockets ready
-		proxy = zmq4.NewProxy(ctx, front, back, capt)
+		proxy, _ = zmq4.NewProxy(ctx, front, back, capt)
```
we should probably check for that error.

stitchinthyme replied:
Done.

sbinet reviewed (commented):
one last wrinkle to iron out.

could you also send another PR against [go-zeromq/license](https://github.com/go-zeromq/license) adding yourself to the `AUTHORS` and/or `CONTRIBUTORS` files?

thanks again.

sbinet commented on proxy_test.go:
```diff
@@ -260,7 +260,10 @@ func TestProxy(t *testing.T) {
+		}
 		t.Logf("proxy ready")
 		wg2.Done()
 		err := proxy.Run()
```
this will probably fail to compile.
this shouldn't:
```suggestion
		err = proxy.Run()
```

stitchinthyme replied:
Added forward declaration of "err", changed := to = on proxy.Run() line.

stitchinthyme commented:
Sorry, didn't notice there was another call to NewProxy(). Let's hope this
time's the charm. :-)

On Mon, Sep 27, 2021 at 12:30 PM Sebastien Binet ***@***.***>
wrote:

> thanks for the update.
> it seems the tests fail to compile:
>
> # github.com/go-zeromq/zmq4_test [github.com/go-zeromq/zmq4.test]
> Error: ./proxy_test.go:321:24: multiple-value zmq4.NewProxy() in single-value context
> FAIL	github.com/go-zeromq/zmq4 [build failed]
> FAIL
> Error: Process completed with exit code 2.
>
> we're nearly there :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-zeromq/zmq4/pull/112#issuecomment-928051505>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC25E3HKJXI42VZ4EMPWEYDUECL2ZANCNFSM5D6JXQ4A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

sbinet commented:
two remaining issues:

```
=== RUN   TestTopics
--- FAIL: TestTopics (0.50s)
panic: interface conversion: *zmq4.pubSocket is not zmq4.Topics: missing method Topics [recovered]
	panic: interface conversion: *zmq4.pubSocket is not zmq4.Topics: missing method Topics
```

and the linter detected:
```
Error: func `(*Conn).subscribed` is unused (unused)
Error: func `(*Conn).subscribe` is unused (unused)
```

stitchinthyme commented:
The stuff in conn.go was me trying something - wasn't intending to commit
it. Anyway, removed it and added Topics to pub and xsub.

On Mon, Sep 27, 2021 at 12:43 PM Sebastien Binet ***@***.***>
wrote:

> two remaining issues:
>
> === RUN   TestTopics
> --- FAIL: TestTopics (0.50s)
> panic: interface conversion: *zmq4.pubSocket is not zmq4.Topics: missing method Topics [recovered]
> 	panic: interface conversion: *zmq4.pubSocket is not zmq4.Topics: missing method Topics
>
> and the linter detected:
>
> Error: func `(*Conn).subscribed` is unused (unused)
> Error: func `(*Conn).subscribe` is unused (unused)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-zeromq/zmq4/pull/112#issuecomment-928060937>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC25E3AM46CJ5K3CYFOX76TUECNMRANCNFSM5D6JXQ4A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

stitchinthyme commented:
Sigh...copypasta gets me every time. :-/

stitchinthyme commented:
So I removed "topics" from the Conn object because it doesn't look like it's needed -- the sub and xsub sockets have their own topics. But socket.topics() looks at that...so just trying to figure out if there really needs to be a topics field in Conn.

stitchinthyme commented:
Also, is there a way I can run this lint check?

stitchinthyme commented:
"go test" fails for me:

# zmq4_test [zmq4.test]
./proxy_test.go:24:47: undefined: zmq4.Devnull
./proxy_test.go:25:47: undefined: zmq4.Devnull
./proxy_test.go:26:47: undefined: zmq4.Devnull
./proxy_test.go:27:47: undefined: zmq4.Devnull
./proxy_test.go:28:47: undefined: zmq4.Devnull
./proxy_test.go:29:47: undefined: zmq4.Devnull
./proxy_test.go:291:47: undefined: zmq4.Devnull
./proxy_test.go:292:47: undefined: zmq4.Devnull
./proxy_test.go:293:47: undefined: zmq4.Devnull
./proxy_test.go:294:47: undefined: zmq4.Devnull
./proxy_test.go:294:47: too many errors
FAIL zmq4 [build failed]
FAIL

On Tue, Sep 28, 2021 at 4:57 AM Sebastien Binet ***@***.***>
wrote:

> the lint stuff is a bit cumbersome to run locally.
>
> but running go test -v ./... should catch most of what the linter
> complains about.
>
> originally, (x)pub and (x)sub would both know the topics they were
> (resp.) publishing or subscribing via this socket.topics.
> with the new scheme, it seems to me only (x)sub know.
> so... perhaps we should just drop Conn.topics and socket.topics
> altogether if not knowing/implementing Topics for (x)pub is alright.
>
> I don't remember off-hand whether a (x)pub socket is expected to know the
> topics it is publishing.
> if it's alright, then we can drop that.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-zeromq/zmq4/pull/112#issuecomment-928993521>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC25E3D5APJSBPRP2S7FEL3UEF7ORANCNFSM5D6JXQ4A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

sbinet commented:
that's quite strange:

```
$> cat zall_test.go 
// Copyright 2020 The go-zeromq Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package zmq4

import (
	"io/ioutil"
	"log"
)

var (
	Devnull = log.New(ioutil.Discard, "zmq4: ", 0)
)
```

what version of Go are you using?

stitchinthyme commented:
$ go version
go version go1.17.1 darwin/amd64

On Tue, Sep 28, 2021 at 3:45 PM Sebastien Binet ***@***.***>
wrote:

> that's quite strange:
>
> $> cat zall_test.go
> // Copyright 2020 The go-zeromq Authors. All rights reserved.
> // Use of this source code is governed by a BSD-style
> // license that can be found in the LICENSE file.
>
> package zmq4
>
> import (
> 	"io/ioutil"
> 	"log"
> )
>
> var (
> 	Devnull = log.New(ioutil.Discard, "zmq4: ", 0)
> )
>
> what version of Go are you using?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-zeromq/zmq4/pull/112#issuecomment-929570361>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC25E3CIXUB3ZY5DHLWVFRDUEILPLANCNFSM5D6JXQ4A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

stitchinthyme commented:
Okay, so the tests pass now. The latest changes:

- Removed topics from main socket type.
- Made pub and xpub Topics() functions return an empty list since those
socket types don't have subscriptions.
- Fixed xsub Topics() to return its topics list (it was returning the one
from the main socket).
- Removed topics tests for pub and xpub sockets in pub/sub and xpub/xsub
tests.
- Added topics test for sub socket in xpub/xsub test.

Crossing fingers that lint likes this one....

On Tue, Sep 28, 2021 at 4:57 AM Sebastien Binet ***@***.***>
wrote:

> the lint stuff is a bit cumbersome to run locally.
>
> but running go test -v ./... should catch most of what the linter
> complains about.
>
> originally, (x)pub and (x)sub would both know the topics they were
> (resp.) publishing or subscribing via this socket.topics.
> with the new scheme, it seems to me only (x)sub know.
> so... perhaps we should just drop Conn.topics and socket.topics
> altogether if not knowing/implementing Topics for (x)pub is alright.
>
> I don't remember off-hand whether a (x)pub socket is expected to know the
> topics it is publishing.
> if it's alright, then we can drop that.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-zeromq/zmq4/pull/112#issuecomment-928993521>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC25E3D5APJSBPRP2S7FEL3UEF7ORANCNFSM5D6JXQ4A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

stitchinthyme commented:
I totally do not understand this line (which is what seems to be a problem for the build):

got := sub.(zmq4.Topics).Topics()

(zmq4_xpubsub_test.go:147)

sbinet commented:
this does an interface-cast: it "casts" the `sub` variable to the `zmq4.Topics` interface and then call the `Topics()` method of that interface.

the pure-Go `sub` sockets do implement the `Topics` interface.
but the shims to the C++ ones do not.

you seem to have moved that code from the `TestTopics` func (that's only run on pure-Go sockets) to the `TestXPubSub` one, that does exercize the Go/Go, Go/C++, C++/Go and C++/C++ cases.

I guess replacing that snippet of code from `TextXPubSub` with:
```go
got, ok := sub.(zmq4.Topics).Topics() // C++ sockets do not implement the zmq4.Topics interface.
want := []string{topics[isub]}
if ok && !reflect.DeepEqual(got, want) {
	t.Fatalf("Missing or wrong topics.\ngot= %q\nwant=%q", got, want)
}
```

stitchinthyme commented:
Might need something like this instead:

t, ok := sub.(zmq4.Topics) // C++ sockets do not implement the
zmq4.Topics interface.want := []string{topics[isub]}if ok {

        if got := t.Topics(); !reflect.DeepEqual(got, want) {
	        t.Fatalf("Missing or wrong topics.\ngot= %q\nwant=%q", got, want)
        }
}


On Wed, Sep 29, 2021 at 11:27 AM Sebastien Binet ***@***.***>
wrote:

> this does an interface-cast: it "casts" the sub variable to the
> zmq4.Topics interface and then call the Topics() method of that interface.
>
> the pure-Go sub sockets do implement the Topics interface.
> but the shims to the C++ ones do not.
>
> you seem to have moved that code from the TestTopics func (that's only
> run on pure-Go sockets) to the TestXPubSub one, that does exercize the
> Go/Go, Go/C++, C++/Go and C++/C++ cases.
>
> I guess replacing that snippet of code from TextXPubSub with:
>
> got, ok := sub.(zmq4.Topics).Topics() // C++ sockets do not implement the zmq4.Topics interface.want := []string{topics[isub]}if ok && !reflect.DeepEqual(got, want) {
> 	t.Fatalf("Missing or wrong topics.\ngot= %q\nwant=%q", got, want)
> }
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-zeromq/zmq4/pull/112#issuecomment-930284791>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC25E3ALZ3UPUITH27WB373UEMV4VANCNFSM5D6JXQ4A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

stitchinthyme commented:
Ok, I have no idea why the build is still failing or what the "Codecov Report" means.

sbinet commented:
It seems one of the tests deadlocks, but only with 1.17...

Oh well...

I'll have a look next week: I am off the grid starting now until Wednesday next week.

In the meantime, could send another pull request against the 'license' repository, adding yourself to the 'AUTHORS' and/or 'CONTRIBUTORS' ?
When I am back I'll merge the 2 PRs (after having fixed the locked test, unless you beat me to it).

Apologies for the bumpy ride and delay(s)

stitchinthyme commented:
How likely am I to be inviting (more) spam by adding my email?

On Wed, Sep 29, 2021 at 5:04 PM Sebastien Binet ***@***.***>
wrote:

> It seems one of the tests deadlocks, but only with 1.17...
>
> Oh well...
>
> I'll have a look next week: I am off the grid starting now until Wednesday
> next week.
>
> In the meantime, could send another pull request against the 'license'
> repository, adding yourself to the 'AUTHORS' and/or 'CONTRIBUTORS' ?
> When I am back I'll merge the 2 PRs (after having fixed the locked test,
> unless you beat me to it).
>
> Apologies for the bumpy ride and delay(s)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-zeromq/zmq4/pull/112#issuecomment-930542491>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC25E3CQ3B2QLTPTT3XOQX3UEN5PBANCNFSM5D6JXQ4A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

sbinet commented:
ok, so...

> How likely am I to be inviting (more) spam by adding my email?

well, I haven't seen any uptick of spam in my inbox although I have my address in a few `CONTRIBUTORS` files on GitHub (`golang`, `gonum`, ...).

if you prefer, you could just put your name + GitHub handle.
